fn/context_guard.go

```diff
@@ -185,31 +186,27 @@ func (g *ContextGuard) ctxQuitUnsafe(ctx context.Context,
 	cancel = g.addCancelFnUnsafe(cancel)
 
 	g.wg.Add(1)
-	go func() {
-		defer cancel()
-		defer g.wg.Done()
 
-		select {
-		case <-g.quit:
-
-		case <-ctx.Done():
-		}
-	}()
+	// We don't have to wait on g.quit here: g.quit can be closed only in
+	// Quit method, which also closes the context we are waiting for.
+	context.AfterFunc(ctx, func() {
```

Comment:
This doesn't really remove any overhead (since `AterContext()` calls `propagateCancel()` which creates the same "wait for context cancel" goroutine. But it's much simpler to read, so nice optimization in any case.
Also nice catch that with the `g.quit` now being un-exported, we can skip that code path as well.

Reply:
Ah, right, that makes sense. Should've looked at the code in `propagateCancel` more closely I guess :see_no_evil: Thanks for the clarification! And the additional unit test.

Reply:
Thanks for this PR! @starius  I understand the core change is to use `context.AfterFunc` instead of a goroutine for context cancellation handling, which should reduce overhead. 

> Without this optimization, the test would initiate 4000 new goroutines.

I'm trying to understand what you mean by this.
Are you referring to starting a goroutine for each context created here (in the test)?

```go

// Create 1000 contexts of each type.
	for i := 0; i < 1000; i++ {
		_, _ = g.Create(ctx)
		_, _ = g.Create(ctx, WithBlockingCG())
		_, _ = g.Create(ctx, WithTimeoutCG())
		_, _ = g.Create(ctx, WithBlockingCG(), WithTimeoutCG())
	}
```

Additionally, the addition of the test is great. Overall, this looks like a solid improvement.

Reply:
> Are you referring to starting a goroutine for each context created here (in the test)?

Yes.